src/zac/core/views/processes.py

```
@@ -364,8 +364,9 @@ def validate_state(self, encoded_state: str) -> dict:
         return state
 
     def complete_task(self) -> HttpResponseRedirect:
-        task = self._get_task()
-        complete_task(task.id, {})
+        task = self._get_task(refresh=True)
+        if not task.historical:
+            complete_task(task.id, {})
```

This is the actual fix - by refreshing as close to the `complete_task` call, we minimize the risk of race conditions. It does not completely eliminate them, but I'd rather see how far we get without wrapping this in all kinds of defensive programming constructions.

If we know that the task is fetched from the history, this implies that it was completed already (from the notifications, in this case).

---

Why not silently catch the exception here if it's already completed and redirect to a different page saying the task was already completed?

---

Because it's not a user mistake - for the user this is silent. The task was already completed in another way, and they just got automatically redirected to this page, which redirects the user back to the detail page. This is not an explicit user action where the user clicks the button "complete task" - it's entirely automated.

So, I want to behaviour to be the same independent from what happens first: either the notification is handled first, or the user redirect is handled first (which corresponds by going to Kownsl from e-mail vs. going to Kownsl from the ZAC tasks)